docker/docker_client.go

```diff
@@ -221,10 +221,93 @@ func CheckAuth(ctx context.Context, sCtx *types.SystemContext, username, passwor
 	}
 }
 
+// V2Data holds the results returned by the /v2/_catalog endpoint
+type V2Data struct {
+	Repositories []string `json:"repositories"`
+}
+
+// V1Data holds the results returned by the /v1/search endpoints
+type V1Data struct {
+	Results []Results `json:"results"`
+}
+
+// Results holds the information of each matching image returned by the v1 endpoint
+type Results struct {
+	Name        string `json:"name"`
+	Description string `json:"description"`
+	StarCount   int    `json:"star_count"`
+	IsTrusted   bool   `json:"is_trusted"`
+	IsAutomated bool   `json:"is_automated"`
+	IsOfficial  bool   `json:"is_official"`
+}
+
+// SearchRegistries queries a list of registries for a matching image
+func SearchRegistries(ctx context.Context, sCtx *types.SystemContext, registries []string, image, limit string) (map[string][]Results, error) {
+	searchResults := make(map[string][]Results)
+	for _, reg := range registries {
+		path := "/v2/_catalog"
+		url := reg
+		if strings.Contains(reg, "docker") {
+			url = "index.docker.io"
+		}
+		logrus.Debugf("pinging v2 endpoint\n")
+		v2Data, _, err := getRegistryData(ctx, sCtx, url, path, true)
+		if err != nil {
+			path = "/v1/search?q=" + image + "&n=" + limit
+			logrus.Debugf("pinging v1 endpoint\n")
+			_, v1Data, err := getRegistryData(ctx, sCtx, url, path, false)
+			if err != nil {
+				fmt.Printf("couldn't search registry %q: %v", reg, err)
+				continue
+			}
+			searchResults[reg] = v1Data.Results
+			continue
+		}
+		results := []Results{}
+		for _, repo := range v2Data.Repositories {
+			if strings.Contains(repo, image) {
+				res := Results{
+					Name: repo,
+				}
+				results = append(results, res)
+			}
+		}
+		searchResults[reg] = results
+	}
+	return searchResults, nil
+}
+
+// getREgistryData talks to either the v2 or v1 endpoint to get the results of the search query
+func getRegistryData(ctx context.Context, sCtx *types.SystemContext, registry, path string, isV2 bool) (*V2Data, *V1Data, error) {
+	v2Data := &V2Data{}
+	v1Data := &V1Data{}
+	newLoginClient, err := newDockerClientWithDetails(sCtx, registry, "", "", "", nil, "")
+	if err != nil {
+		return nil, nil, errors.Wrapf(err, "error creating new docker client")
+	}
+	resp, err := newLoginClient.makeRequest(ctx, "GET", path, nil, nil, true)
+	if err != nil {
+		return nil, nil, err
+	}
+	if resp.StatusCode != http.StatusOK {
+		return nil, nil, errors.Errorf("error getting response from endpoint")
+	}
+	if isV2 {
```

Comment:
The way this is split, with the URL path hard-coded in the caller and the data format hard-coded in the callee, is weird.  Maybe instead the front part of `getRegistryData` should become a “GET + check for StatusOK” helper?

Why do the V1 and V2 requests use a separate `dockerClient` anyway?

Reply:
restructured it quite a bit. Should be fixed now.